File: examples/typescript/test/workflow/subscription/subscription-billing.test.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import assert from "node:assert/strict";
import test from "node:test";

import { MyDependencyService } from "../../../src/workflow/my-dependency-service.js";
import type { Customer } from "../../../src/workflow/subscription/models.js";
import { SubscriptionBilling } from "../../../src/workflow/subscription/subscription-billing.js";

const TEST_CUSTOMER: Customer = {
  firstName: "Quanzheng",
  lastName: "Long",
  id: "123",
  email: "qlong.seattle@gmail.com",
  subscription: {
    trialPeriodMs: 2_000,
    billingPeriodMs: 1_000,
    maxBillingPeriods: 10,
    billingPeriodCharge: 100,
  },
};

class RecordingService extends MyDependencyService {
  public readonly emails: Array<{
    recipient: string;
    subject: string;
    content: string;
  }> = [];
  public readonly charges: Array<{
    email: string;
    customerId: string;
    amount: number;
  }> = [];

  public override sendEmail(recipient: string, subject: string, content: string): void {
    this.emails.push({ recipient, subject, content });
  }

  public override chargeUser(email: string, customerId: string, amount: number): void {
    this.charges.push({ email, customerId, amount });
  }
}

test("sendWelcomeEmail", () => {
  const service = new RecordingService();
  SubscriptionBilling.sendWelcomeEmail(TEST_CUSTOMER, service);
  assert.deepEqual(service.emails, [
    {
      recipient: TEST_CUSTOMER.email,
      subject: "welcome email",
      content: "hello content",
    },
  ]);
  assert.equal(SubscriptionBilling.trialPeriod(TEST_CUSTOMER), 2_000);
});

test("isSubscriptionOver", () => {
  assert.equal(SubscriptionBilling.isSubscriptionOver(TEST_CUSTOMER, 0), false);
  assert.equal(
    SubscriptionBilling.isSubscriptionOver(
      TEST_CUSTOMER,
      TEST_CUSTOMER.subscription.maxBillingPeriods,
    ),
    true,
  );
  assert.equal(SubscriptionBilling.billingPeriod(TEST_CUSTOMER), 1_000);
});

test("chargeCurrentPeriod", () => {
  const service = new RecordingService();
  SubscriptionBilling.chargeCurrentPeriod(TEST_CUSTOMER, service);
  assert.deepEqual(service.charges, [
    {
      email: TEST_CUSTOMER.email,
      customerId: TEST_CUSTOMER.id,
      amount: 100,
    },
  ]);
  assert.equal(service.emails.length, 0);
});

test("sendSubscriptionOverEmail", () => {
  const service = new RecordingService();
  SubscriptionBilling.sendSubscriptionOverEmail(TEST_CUSTOMER, service);
  assert.equal(service.charges.length, 0);
  assert.deepEqual(service.emails, [
    {
      recipient: TEST_CUSTOMER.email,
      subject: "subscription over",
      content: "hello content",
    },
  ]);
});

test("sendCanceledEmail", () => {
  const service = new RecordingService();
  SubscriptionBilling.sendCanceledEmail(TEST_CUSTOMER, service);
  assert.deepEqual(service.emails, [
    {
      recipient: TEST_CUSTOMER.email,
      subject: "subscription canceled",
      content: "hello content",
    },
  ]);
});

test("applyChargeAmount", () => {
  const customer: Customer = {
    ...TEST_CUSTOMER,
    subscription: { ...TEST_CUSTOMER.subscription },
  };
  SubscriptionBilling.applyChargeAmount(customer, 200);
  assert.equal(customer.subscription.billingPeriodCharge, 200);
  assert.equal(SubscriptionBilling.requireSingleChargeAmount([200]), 200);
});

test("requireSingleChargeAmountRejectsUnexpectedResults", () => {
  assert.throws(() => SubscriptionBilling.requireSingleChargeAmount(null));
  assert.throws(() => SubscriptionBilling.requireSingleChargeAmount([]));
  assert.throws(() => SubscriptionBilling.requireSingleChargeAmount([100, 200]));
});

File: examples/java/src/test/java/io/superdurable/dex/products/subscription/SubscriptionBillingTest.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.superdurable.dex.products.subscription;

import io.superdurable.dex.shared.MyDependencyService;
import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class SubscriptionBillingTest {
    private static final Customer TEST_CUSTOMER = new Customer(
            "Quanzheng",
            "Long",
            "123",
            "qlong.seattle@gmail.com",
            new Subscription(
                    Duration.ofSeconds(2),
                    Duration.ofSeconds(1),
                    10,
                    100));

    @Test
    void sendWelcomeEmail() {
        final RecordingService service = new RecordingService();

        SubscriptionBilling.sendWelcomeEmail(TEST_CUSTOMER, service);

        assertEquals(
                Collections.singletonList(new RecordedEmail(
                        TEST_CUSTOMER.email,
                        "welcome email",
                        "hello content")),
                service.emails);
        assertEquals(Duration.ofSeconds(2), SubscriptionBilling.trialPeriod(TEST_CUSTOMER));
    }

    @Test
    void isSubscriptionOver() {
        assertFalse(SubscriptionBilling.isSubscriptionOver(TEST_CUSTOMER, 0));
        assertTrue(SubscriptionBilling.isSubscriptionOver(
                TEST_CUSTOMER,
                TEST_CUSTOMER.subscription.maxBillingPeriods));
        assertEquals(Duration.ofSeconds(1), SubscriptionBilling.billingPeriod(TEST_CUSTOMER));
    }

    @Test
    void chargeCurrentPeriod() {
        final RecordingService service = new RecordingService();

        SubscriptionBilling.chargeCurrentPeriod(TEST_CUSTOMER, service);

        assertEquals(
                Collections.singletonList(new RecordedCharge(
                        TEST_CUSTOMER.email,
                        TEST_CUSTOMER.id,
                        100)),
                service.charges);
        assertTrue(service.emails.isEmpty());
    }

    @Test
    void sendSubscriptionOverEmail() {
        final RecordingService service = new RecordingService();

        SubscriptionBilling.sendSubscriptionOverEmail(TEST_CUSTOMER, service);

        assertTrue(service.charges.isEmpty());
        assertEquals(
                Collections.singletonList(new RecordedEmail(
                        TEST_CUSTOMER.email,
                        "subscription over",
                        "hello content")),
                service.emails);
    }

    @Test
    void sendCanceledEmail() {
        final RecordingService service = new RecordingService();

        SubscriptionBilling.sendCanceledEmail(TEST_CUSTOMER, service);

        assertEquals(
                Collections.singletonList(new RecordedEmail(
                        TEST_CUSTOMER.email,
                        "subscription canceled",
                        "hello content")),
                service.emails);
    }

    @Test
    void applyChargeAmount() {
        final Customer customer = new Customer(
                TEST_CUSTOMER.firstName,
                TEST_CUSTOMER.lastName,
                TEST_CUSTOMER.id,
                TEST_CUSTOMER.email,
                new Subscription(
                        TEST_CUSTOMER.subscription.trialPeriod,
                        TEST_CUSTOMER.subscription.billingPeriod,
                        TEST_CUSTOMER.subscription.maxBillingPeriods,
                        TEST_CUSTOMER.subscription.billingPeriodCharge));

        SubscriptionBilling.applyChargeAmount(customer, 200);

        assertEquals(200, customer.subscription.billingPeriodCharge);
        assertEquals(200, SubscriptionBilling.requireSingleChargeAmount(
                Collections.singletonList(200)));
    }

    @Test
    void requireSingleChargeAmountRejectsUnexpectedResults() {
        assertThrows(
                IllegalStateException.class,
                () -> SubscriptionBilling.requireSingleChargeAmount(null));
        assertThrows(
                IllegalStateException.class,
                () -> SubscriptionBilling.requireSingleChargeAmount(Collections.emptyList()));
        assertThrows(
                IllegalStateException.class,
                () -> SubscriptionBilling.requireSingleChargeAmount(Arrays.asList(100, 200)));
    }

    private static final class RecordedEmail {
        private final String recipient;
        private final String subject;
        private final String content;

        private RecordedEmail(
                final String recipient, final String subject, final String content) {
            this.recipient = recipient;
            this.subject = subject;
            this.content = content;
        }

        @Override
        public boolean equals(final Object other) {
            if (!(other instanceof RecordedEmail)) {
                return false;
            }
            final RecordedEmail that = (RecordedEmail) other;
            return recipient.equals(that.recipient)
                    && subject.equals(that.subject)
                    && content.equals(that.content);
        }

        @Override
        public int hashCode() {
            return Arrays.hashCode(new Object[] {recipient, subject, content});
        }
    }

    private static final class RecordedCharge {
        private final String email;
        private final String customerId;
        private final int amount;

        private RecordedCharge(final String email, final String customerId, final int amount) {
            this.email = email;
            this.customerId = customerId;
            this.amount = amount;
        }

        @Override
        public boolean equals(final Object other) {
            if (!(other instanceof RecordedCharge)) {
                return false;
            }
            final RecordedCharge that = (RecordedCharge) other;
            return email.equals(that.email)
                    && customerId.equals(that.customerId)
                    && amount == that.amount;
        }

        @Override
        public int hashCode() {
            return Arrays.hashCode(new Object[] {email, customerId, amount});
        }
    }

    private static final class RecordingService extends MyDependencyService {
        private final List<RecordedEmail> emails = new ArrayList<RecordedEmail>();
        private final List<RecordedCharge> charges = new ArrayList<RecordedCharge>();

        @Override
        public void sendEmail(
                final String recipient, final String subject, final String content) {
            emails.add(new RecordedEmail(recipient, subject, content));
        }

        @Override
        public void chargeUser(final String email, final String customerId, final int amount) {
            charges.add(new RecordedCharge(email, customerId, amount));
        }
    }
}
